src/Features/Core/Portable/SemanticSearch/AbstractSemanticSearchService.cs

```diff
@@ -0,0 +1,436 @@
+                debugInformationFormat: DebugInformationFormat.PortablePdb,
+                instrumentationKinds: [InstrumentationKind.StackOverflowProbing, InstrumentationKind.ModuleCancellation]);
+
+            using var peStream = new MemoryStream();
+            using var pdbStream = new MemoryStream();
+
+            var emitDifferenceTimer = SharedStopwatch.StartNew();
+            var emitResult = queryCompilation.Emit(peStream, pdbStream, options: emitOptions, cancellationToken: cancellationToken);
+            var emitTime = emitDifferenceTimer.Elapsed;
+
+            var executionTime = TimeSpan.Zero;
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // complete compilation progress item:
+            remainingProgressItemCount--;
+            await observer.ItemsCompletedAsync(1, cancellationToken).ConfigureAwait(false);
+
+            if (!emitResult.Success)
+            {
+                foreach (var diagnostic in emitResult.Diagnostics)
+                {
+                    if (diagnostic.Severity == DiagnosticSeverity.Error)
+                    {
+                        traceSource.TraceInformation($"Semantic search query compilation failed: {diagnostic}");
+                    }
+                }
+
+                await observer.OnCompilationFailureAsync(
+                    emitResult.Diagnostics.SelectAsArray(
+                        d => d.Severity == DiagnosticSeverity.Error,
+                        d => new QueryCompilationError(d.Id, d.GetMessage(), (d.Location.SourceTree == queryTree) ? d.Location.SourceSpan : default)),
+                    cancellationToken).ConfigureAwait(false);
+
+                return CreateResult(FeaturesResources.Semantic_search_query_failed_to_compile);
+            }
+
+            peStream.Position = 0;
+            pdbStream.Position = 0;
+            var loadContext = new LoadContext();
+            try
+            {
+                var queryAssembly = loadContext.LoadFromStream(peStream, pdbStream);
+
+                var pidType = queryAssembly.GetType("<PrivateImplementationDetails>", throwOnError: true);
+                Contract.ThrowIfNull(pidType);
+                var moduleCancellationTokenField = pidType.GetField("ModuleCancellationToken", BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Static);
+                Contract.ThrowIfNull(moduleCancellationTokenField);
+                moduleCancellationTokenField.SetValue(null, cancellationToken);
+
+                if (!TryGetFindMethod(queryAssembly, out var findMethod, out var errorMessage, out var errorMessageArgs))
+                {
+                    traceSource.TraceInformation($"Semantic search failed: {errorMessage}");
+                    return CreateResult(errorMessage, errorMessageArgs);
+                }
+
+                var executionTimeStopWatch = new Stopwatch();
+
+                foreach (var project in solution.Projects)
```

Comment:
Could we do the search in parallel?

Reply:
We could but there is a chance the query defines global state. So there is a tradeoff between simplicity and performance here.

Reply:
Maybe leave some comments here to explain this. Later we might want to let user opt-in to run the query in parallel in some way

Reply:
I'll follow up on this.